In [17]:
import numpy as np
import pandas as pd

from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

import tensorflow as tf
from tensorflow.keras.utils import to_categorical
from tensorflow.keras.layers import Dense
from tensorflow.keras.models import Sequential

In [18]:
data = load_iris()

X, y = data.data, data.target

# Sprawdzam ile cech ma iris
print(X.shape) 

(150, 4)


In [19]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, stratify=y, random_state=42)

In [20]:
scaler = StandardScaler()

X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

In [21]:
# Zmiana klas na categorical, czyli w formie, np. (0,0,1)
y_train = tf.keras.utils.to_categorical(y_train)
y_test = tf.keras.utils.to_categorical(y_test)

In [23]:
# Dałem shape do pierwszej warstwy (za "relu"), ale wyczytałem, że określa się liczbę cech przed warstwami. Wcześniej wyskakiwało mi ostrzeżenie.

model = Sequential([
    tf.keras.Input(shape=(4,)),
    Dense(16, activation='relu'),
    Dense(8, activation='relu'),
    Dense(3, activation='softmax')
])

In [24]:
model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

In [25]:
learn = model.fit(X_train, y_train, epochs=100, validation_split=0.2, batch_size=8)

Epoch 1/100
12/12 ━━━━━━━━━━━━━━━━━━━━ 1s 11ms/step - accuracy: 0.3125 - loss: 1.3028 - val_accuracy: 0.4167 - val_loss: 1.1862
Epoch 2/100
12/12 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.3125 - loss: 1.2375 - val_accuracy: 0.4167 - val_loss: 1.1390
Epoch 3/100
12/12 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.3125 - loss: 1.1837 - val_accuracy: 0.4583 - val_loss: 1.1023
Epoch 4/100
12/12 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.3542 - loss: 1.1449 - val_accuracy: 0.5417 - val_loss: 1.0717
Epoch 5/100
12/12 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.4167 - loss: 1.1115 - val_accuracy: 0.6250 - val_loss: 1.0506
Epoch 6/100
12/12 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.4479 - loss: 1.0845 - val_accuracy: 0.6250 - val_loss: 1.0332
Epoch 7/100
12/12 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.4688 - loss: 1.0575 - val_accuracy: 0.6250 - val_loss: 1.0124
Epoch 8/100
12/12 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - accuracy: 0.5000 - loss: 1.0301 - val_accuracy: 0.6667 - 

In [26]:
loss, accuracy = model.evaluate(X_test, y_test)

print("Accuracy:", accuracy)

1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 93ms/step - accuracy: 0.9333 - loss: 0.1310
Accuracy: 0.9333333373069763


In [28]:
prediction = model.predict(X_test)

print(prediction[:10])

1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 19ms/step
[[9.9927145e-01 7.1582280e-04 1.2702903e-05]
 [4.4944414e-04 1.4709239e-01 8.5245812e-01]
 [1.3885148e-02 9.8256713e-01 3.5476584e-03]
 [5.0276243e-03 9.9277711e-01 2.1952458e-03]
 [9.9973279e-01 2.5624197e-04 1.1005396e-05]
 [9.4413571e-03 6.9997609e-01 2.9058254e-01]
 [9.9983788e-01 1.4639254e-04 1.5684560e-05]
 [9.9565059e-01 4.1379328e-03 2.1154605e-04]
 [4.2592601e-05 3.7280023e-02 9.6267736e-01]
 [5.8968938e-03 7.9243922e-01 2.0166384e-01]]


Sieć neuronowa wyuczyła model osiągając skuteczność (trafność) 93%. Co jest dobrym wynikiem.